### PR TITLE
Handle QVariant::Type in PyQt6 builds

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR adds compatibility for calling PyQGIS methods which require QVariant::Type arguments (or which return QVariant::Type values).

QVariant::Type does not exist in PyQt6 as its been deprecated in Qt6 and replaced with QMetaType::Type.

In order to avoid breaking PyQGIS API, we don't want to change all our functions to use QMetaType::Type instead of QVariant::Type (that can wait till QGIS 4.0). So instead we leave the c++/Qt 5 signatures as QVariant::Type, but accept QMetaType::Type values for these functions under Qt 6 builds. SIP MappedType code has been added to automatically convert from QMetaType::Type <-> QVariant::Type when calling PyQGIS methods or handling their return values.

Additionally, we monkey patch in QVariant::Type values for the PyQt6 builds. Since QVariant.Type doesn't exist in PyQt6, we patch it back in by setting them to their QMetaType.Type equivalents.

This means that code which calls eg:

     field = QgsField('my_field', QVariant.Int)

will work as usual under Qt5, while on Qt6 builds it will also work without any change. (Since QVariant.Int is QMetaType.Type.Int, and the PyQGIS6 sip conversion code will kick in and transparently convert the QMetaType.Type.Int value to QVariant::Int when calling the underlying c++ method)

### CODE REVIEW

**Comment by nyalldawson** [comment]
@troopa81 I managed to get it working (I was incorrectly returning the wrong value in ConvertToTypeCode when sipIsErr == NULL, so sip would never recognise that QMetaType was convertible to QVariant::Type).

**Comment by nyalldawson** [comment]
Ping @manisandro , you've also raised questions about this in the past

File: python/PyQt/PyQt/QtCore.py.in [hunk 1]
```diff
@@ -72,5 +72,64 @@ QVariant.__hash__ = __hash__
 
 if (QT_VERSION >= 0x060000):
     NULL = None
+    # monkey patch in QVariant Types
```
Comment (troopa81) [inline python/PyQt/PyQt/QtCore.py.in:75]: Why not relying on the migration script pyqt5_to_pyqt6 to change this one directly in the python script ?

**Review by troopa81 — COMMENTED** [review]
I would have preferred get rid completely of QVariant::Type. It would be better no? or is it a first stage before doing so?

Does it work to pass an integer to these methods ? It seems some plugin were doing this kind of things and it was working in pyqt5 IIRC

**Comment by nyalldawson** [comment]
> I would have preferred get rid completely of QVariant::Type. It would be better no? or is it a first stage before doing so?

It's for pyqt5 compatibility, so that it's straightforward to support both versions without intrusive changes in a plugin. We can reassess for qgis 4.0 in a couple of years 👍

> Does it work to pass an integer to these methods ? It seems some plugin were doing this kind of things and it was working in pyqt5 IIRC

I'll test, but imo that's a coding bug in those plugins. Enums should never rely on a static integer value.

File: python/PyQt/PyQt/QtCore.py.in (hunk 1 above)
Comment (troopa81) [reply python/PyQt/PyQt/QtCore.py.in:75]: Oh yes, I forgot

**Comment by troopa81** [comment]
> It's for pyqt5 compatibility, so that it's straightforward to support both versions without intrusive changes in a plugin. We can reassess for qgis 4.0 in a couple of years 👍

What if we migrate QVariant::Type to QMetaType::Type and then make the binding (with sipify maybe) still support QVariant::Type. This way, we migrate and we are not breaking compatbility.

I didn't test it and I don't know if it's really possible, just an idea.

**Comment by nyalldawson** [comment]
I'd prefer the minimal approach for now, and defer larger changes till 4.0. changing QVariant::Type throughout the c++ codebase will be a massive amount of churn

**Comment by troopa81** [comment]
> @troopa81 Thinking about it overnight, the other very compelling reason to leave as QVariant::Type is that by moving to QMetaType we'd break compatibility for plugins with all QGIS versions < 3.34/3.36 (ie the unmaintained finalised releases).

Yes, you right we have to keep the QVariant.Type in plugins to keep compatbility, but we can still migrate the API as long that we support both type. in theory...

But I agree with the proposed approach for now

**Review by troopa81 — APPROVED** [review]

**Pull request merged**